libethash-cuda/CUDAMiner.cpp

```diff
@@ -170,7 +171,7 @@ void CUDAMiner::workLoop()
 				if (s_dagLoadIndex >= s_numInstances && s_dagInHostMemory)
 				{
 					// all devices have loaded DAG, we can free now
-					delete[] s_dagInHostMemory;
+					delete [] (uint8_t *)s_dagInHostMemory;
```

Comment:
I believe there is something wrong with `s_dagInHostMemory` being `volatile`. This way we will only hide the issue. Can you try to remove `volatile`?

Reply:
I think the virtual attribute is necessary, specifically in CUDAMiner.cpp around line 140, where:

>// wait until DAG is created on selected device
>while (s_dagInHostMemory == NULL) {
>this_thread::sleep_for(chrono::seconds(1));
>}

Without the volatile attribute, the compiler might optimize out the load of s_dagInHostMemory it considers invariant over the loop. The volatile attribute forces a load with every iteration.

The compiler complains about the delete, as it stands, because we are attempting to delete an array of void elements. Sine the elements are void, technically it can't know the size of the array. In reality, the run-time doesn't care since it has it's own way of knowing the length of the memory block. This is certainly true for the GCC and MSVC runtimes.

I simply changed the delete to refer to an array of uint8_t. Though it makes no difference, that is the way it was allocated in ethash_cuda_miner::init.

Reply:
If you use `volatile` for thread sync, this is wrong. Make the pointer atomic, not the memory volatile.

Reply:
The atomic semantic does not guarantee that the variable load will not be optimized out if the variable is considered invariant over the loop.

Reply:
It does unless you choose a more relaxed memory order than the default one.

Reply:
Ok, I'll verify that (disassemble some generated code) and get back.